libs/grok/src/main/java/org/opensearch/grok/Grok.java

```diff
@@ -125,46 +122,60 @@ private Grok(Map<String, String> patternBank, String grokPattern, boolean namedC
     }
 
     /**
-     * Checks whether patterns reference each other in a circular manner and if so fail with an exception
+     * Entry point to recursively validate the pattern bank for circular dependencies and malformed URLs
+     * via depth-first traversal. This implementation does not
+     */
+    private void validatePatternBank() {
+        for (String patternName : patternBank.keySet()) {
+            validatePatternBank(patternName, new Stack<>());
+        }
+    }
+
+    /**
+     * Checks whether patterns reference each other in a circular manner and if so fail with an exception.
+     * Also checks for malformed pattern definitions and fails with an exception.
      *
      * In a pattern, anything between <code>%{</code> and <code>}</code> or <code>:</code> is considered
      * a reference to another named pattern. This method will navigate to all these named patterns and
      * check for a circular reference.
      */
-    private void forbidCircularReferences(String patternName, List<String> path, String pattern) {
-        if (pattern.contains("%{" + patternName + "}") || pattern.contains("%{" + patternName + ":")) {
+    private void validatePatternBank(String patternName, Stack<String> path) {
+        String pattern = patternBank.get(patternName);
+        boolean isSelfReference = pattern.contains("%{" + patternName + "}") ||
+                                    pattern.contains("%{" + patternName + ":");
+        if (isSelfReference || path.contains(patternName)) {
             String message;
-            if (path.isEmpty()) {
+            if (isSelfReference) {
                 message = "circular reference in pattern [" + patternName + "][" + pattern + "]";
             } else {
-                message = "circular reference in pattern [" + path.remove(path.size() - 1) + "][" + pattern +
```

Comment:
Smells like something you can probably refactor this to be an if/else if/else, or a helper method that throws.